ci: use different payments secrets for separate staging environments

wanlingt opened the pull request:
## Problem
<!-- What problem are you trying to solve? What issue does this close? -->
Previously, all the staging environments shared a single parameter store for `payments` secrets. This meant that we were **unable** to:
1. Use different Stripe accounts for each environment 
(relevant env vars: `PAYMENT_STRIPE_PUBLISHABLE_KEY`, `PAYMENT_STRIPE_SECRET_KEY`, `PAYMENT_STRIPE_CLIENT_ID`)
2. Even within the same Stripe account, use different webhook endpoints for each staging environment as each endpoint has a different signing secret.
(relevant env var: `PAYMENT_STRIPE_WEBHOOK_SECRET`)

Closes #6167, closes #6168

## Solution
<!-- How did you solve the problem? -->
- In AWS, add additional payments [parameter stores](https://ap-southeast-1.console.aws.amazon.com/systems-manager/parameters/?region=ap-southeast-1&tab=Table) for the different staging environments (1 for each - so now there's `staging-payment`, `staging-alt-payment`, `staging-alt2-payment`)
- On the Stripe dashboard, create different accounts for each staging environment
- Rename `ENV_NAME` to `ENV_TYPE` to better describe its function (`prod`, `staging` or `uat`)
- Use a new env var `ENV_NAME` in `.ebextensions/env-file-creation.config`, which determines the name of the parameter store for payments secrets. `ENV_NAME` is taken from `ACTUAL_ENV_NAME` if it's available, otherwise it defaults to`ENV_TYPE`
  - This allows us to have separation of payments secrets. Each staging environment now has its own parameter store, which means that different Stripe accounts and different webhook endpoints can be used.

**Breaking Changes** 
<!-- Does this PR contain any backward incompatible changes? If so, what are they and should there be special considerations for release? -->
- [x] No - this PR is backwards compatible  

## Tests
<!-- What tests should be run to confirm functionality? -->
- [ ] All 3 staging sites (`staging`, `staging-alt`, `staging-alt2`) should be working with payments. To double check that the deployment was successful, retrieve the logs from the EC2 instances. The echo commands can be found in `cfn-init-cmd.log`.

## Deploy Notes
<!-- Notes regarding deployment of the contained body of work.  -->
<!-- These should note any new dependencies, new scripts, etc. -->

**New environment variables**:

- `SSM_ENV_SITE_NAME` : env var in ELB which provides the actual name of the specific environment e.g. `staging-alt`, `staging-alt2`. This is not necessary for `prod`, `staging` and `uat`.

wanlingt committed:
ci: add new ENV_NAME to create .env file script

wanlingt committed:
docs: update env vars used in aws systems manager

wanlingt committed:
fix: update ENV_NAME conditional

LinHuiqing commented:
> Closes https://github.com/opengovsg/FormSG/issues/6167 and https://github.com/opengovsg/FormSG/issues/6168

Btw, for Github to automatically close both issues you need to do something more like "Closes #6167, closes #6168". ([reference](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))

wanlingt commented:
> > Closes #6167 and #6168
> 
> Btw, for Github to automatically close both issues you need to do something more like "Closes #6167, closes #6168". ([reference](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))

ooh thanks! updated the description

LinHuiqing commented on .ebextensions/env-file-creation.config:
```diff
@@ -25,17 +26,27 @@ files:
+
+        echo "Creating ENV_NAME config..."
+        if [ -n "${ACTUAL_ENV_NAME}" ]; then
+            echo "setting ENV_NAME as ACTUAL_ENV_NAME"
```
```suggestion
            echo "setting ENV_NAME as ACTUAL_ENV_NAME (${ACTUAL_ENV_NAME})"
```
Maybe we can be more verbose and `echo` the actual env var to be loaded as well to it'll be easier for us to debug in the future.

wanlingt replied:
makes sense! done in [fe6420c](https://github.com/opengovsg/FormSG/pull/6181/commits/fe6420c5f73b85a4ac72574513819c24cded4d8d)

LinHuiqing commented on .ebextensions/env-file-creation.config:
```diff
@@ -25,17 +26,27 @@ files:
+            echo "setting ENV_NAME as ACTUAL_ENV_NAME"
+            ENV_NAME=${ACTUAL_ENV_NAME}
+        else
+            echo "ACTUAL_ENV_NAME not present, setting ENV_NAME as ENV_TYPE instead..."
```
```suggestion
            echo "ACTUAL_ENV_NAME not present, setting ENV_NAME as ENV_TYPE (${ENV_TYPE}) instead..."
```
Same as above - being more verbose might help in the future

wanlingt replied:
[fe6420c](https://github.com/opengovsg/FormSG/pull/6181/commits/fe6420c5f73b85a4ac72574513819c24cded4d8d)

LinHuiqing commented on .ebextensions/env-file-creation.config:
```diff
@@ -25,17 +26,27 @@ files:
-        aws ssm get-parameter --name "${ENV_NAME}-verified-fields" --with-decryption --region $AWS_REGION | jq -r '.Parameter.Value' >> $TARGET_DIR/.env
-        aws ssm get-parameter --name "${ENV_NAME}-webhook-verified-content" --with-decryption --region $AWS_REGION | jq -r '.Parameter.Value' >> $TARGET_DIR/.env
+
+        echo "Creating ENV_NAME config..."
```
```suggestion
        echo "Creating ENV_NAME config for payment credentials..."
```
Should note here or as a comment that `ENV_NAME` is only used for payment credentials while other credentials use `ENV_TYPE` for the prefix instead

wanlingt replied:
[fe6420c](https://github.com/opengovsg/FormSG/pull/6181/commits/fe6420c5f73b85a4ac72574513819c24cded4d8d)

LinHuiqing commented on .ebextensions/env-file-creation.config:
```diff
@@ -25,17 +26,27 @@ files:
+            ENV_NAME=${ENV_TYPE}
+        fi
+
+        echo "Creating config for ${ENV_TYPE} in ${AWS_REGION}"
```
```suggestion
        echo "Creating config for ${ENV_TYPE}/${ENV_NAME} in ${AWS_REGION}"
```
Might be more helpful if `ENV_NAME` is included here as well!

wanlingt replied:
replaced with `${ENV_SITE_NAME}`, in [fe6420c](https://github.com/opengovsg/FormSG/pull/6181/commits/fe6420c5f73b85a4ac72574513819c24cded4d8d) - so it'll state either prod, staging, staging-alt, staging-alt2, uat

LinHuiqing reviewed (commented):
Small changes to be more verbose in `echo`s! If the code has been tested for all staging envs, should be good to go!

KenLSM commented:
Wondering if `SSM_ACTUAL_ENV_NAME` should be called `SSM_ENV_SITE_NAME`

wanlingt committed:
ref: use ENV_SITE_NAME instead of ACTUAL_ENV_NAME

wanlingt commented:
> Wondering if `SSM_ACTUAL_ENV_NAME` should be called `SSM_ENV_SITE_NAME`

Makes sense to me, changed :) Also changed the IF conditional as discussed - using ENV_SITE_NAME directly instead of creating an additional variable 
[fe6420c](https://github.com/opengovsg/FormSG/pull/6181/commits/fe6420c5f73b85a4ac72574513819c24cded4d8d)

KenLSM commented on docs/DEPLOYMENT_SETUP.md:
```diff
@@ -101,17 +101,20 @@ Store of AWS Service Manager. These groups have names formatted as `<environment
 | `SSM_PREFIX` | String prefix (typically the environment name) for AWS SSM parameter names to create a .env file for FormSG.          |
-| `SECRET_ENV` | String (typically the environment name) to be used in building of AWS Secrets Manager keys in different environments. |
+| `SECRET_ENV` | String (typically the environment name) to be used in building of AWS Secrets Manager keys in different environments. (`staging`, `prod`)|
+| `SSM_ACTUAL_ENV_NAME` | String (the specific environment name) to be used in building of AWS Secrets Manager keys in different environments. Optional. (`staging-alt`, `staging-alt2`)|
```
non-blocking. Should this be `SSM_ENV_SITE_NAME` after the recent name change?

```suggestion
| `SSM_ENV_SITE_NAME` | String (the specific environment name) to be used in building of AWS Secrets Manager keys in different environments. Optional. (`staging-alt`, `staging-alt2`)|
```

wanlingt replied:
thanks for spotting! [9c0da8b](https://github.com/opengovsg/FormSG/pull/6181/commits/9c0da8b4af815242a53a4e3707cd6b23c644f86a)

KenLSM commented on docs/DEPLOYMENT_SETUP.md:
```diff
@@ -101,17 +101,20 @@ Store of AWS Service Manager. These groups have names formatted as `<environment
 The environment for each group is user-defined, and should be specified in the Elastic Beanstalk configuration
 as the environment variable `SSM_PREFIX`.
+The specific environment is user-defined, and should be specified in the Elastic Beanstalk configuration
+as the environment variable `SSM_ACTUAL_ENV_NAME`. This variable is optional.
```
```suggestion
as the environment variable `SSM_ENV_SITE_NAME`. This variable is optional.
```

wanlingt replied:
[9c0da8b](https://github.com/opengovsg/FormSG/pull/6181/commits/9c0da8b4af815242a53a4e3707cd6b23c644f86a)

LinHuiqing reviewed (approved):
lgtm too! 
please try the "Enable auto-merge" button after you've pushed your commits while the checks are running 👀 it should get the PR merged automatically after all required checks have been run 🤞

wanlingt committed:
fix: rename variable

wanlingt committed:
docs: add example for SSM_PREFIX

wanlingt commented:
> lgtm too! please try the "Enable auto-merge" button after you've pushed your commits while the checks are running 👀 it should get the PR merged automatically after all required checks have been run 🤞

it worked!
